services/gundeck/src/Gundeck/Redis.hs

```diff
@@ -0,0 +1,127 @@
+-- the initial connection are used when reconnecting.
+--
+-- Throws 'ConnectError', 'ConnectTimeout', 'ConnectionLostException',
+-- 'PingException', or 'IOException' if retry policy is finite.
+connectRobust ::
+  Logger ->
+  -- | e. g., @exponentialBackoff 50000@
+  RetryPolicy ->
+  -- | action returning a fresh initial 'Connection', e. g., @(connect connInfo)@ or @(connectCluster connInfo)@
+  IO Connection ->
+  IO RobustConnection
+connectRobust l retryStrategy connectLowLevel = do
+  robustConnection <- newEmptyMVar @IO @ReConnection
+  reconnectRedis robustConnection
+  pure robustConnection
+  where
+    reconnectRedis robustConnection = do
+      conn <- connectLowLevel
+
+      Log.info l $ Log.msg (Log.val "lazy connection established, running ping...")
+      -- FUTUREWORK: With ping, we only verify that a single node is running as
+      -- opposed to verifying that all nodes of the cluster are up and running.
+      -- It remains unclear how cluster health can be verified in hedis.
+      void . runRedis conn $ do
+        res <- ping
+        case res of
+          Left r -> throwIO $ PingException r
+          Right _ -> pure ()
+      Log.info l $ Log.msg (Log.val "ping went through")
+
+      reconnectOnce <-
+        once $ -- avoid concurrent attempts to reconnect
+          recovering -- retry connecting, e. g., with exponential back-off
+            retryStrategy
+            [ const $ Catch.Handler (\(e :: ConnectError) -> logEx (Log.err l) e "Redis not in cluster mode" >> pure True),
+              const $ Catch.Handler (\(e :: ConnectTimeout) -> logEx (Log.err l) e "timeout when connecting to Redis" >> pure True),
+              const $ Catch.Handler (\(e :: ConnectionLostException) -> logEx (Log.err l) e "Redis connection lost during request" >> pure True),
+              const $ Catch.Handler (\(e :: PingException) -> logEx (Log.err l) e "pinging Redis failed" >> pure True),
+              const $ Catch.Handler (\(e :: IOException) -> logEx (Log.err l) e "network error when connecting to Redis" >> pure True)
+            ]
+            $ const $
+              reconnectRedis robustConnection
+      let newReConnection = ReConnection {_rrConnection = conn, _rrReconnect = reconnectOnce}
+      unlessM (tryPutMVar robustConnection newReConnection) $
+        void $ swapMVar robustConnection newReConnection
+
+-- | Run a 'Redis' action through a 'RobustConnection'.
+--
+-- Blocks on connection errors as long as the connection is not reestablished.
+-- Without externally enforcing timeouts, this may lead to leaking threads.
+runRobust :: RobustConnection -> Redis a -> IO a
+runRobust mvar action = do
+  robustConnection <- readMVar mvar
+  catches
+    (runRedis (_rrConnection robustConnection) action)
+    [ Handler (\(_ :: ConnectionLostException) -> reconnectRetry robustConnection), -- Redis connection lost during request
+      Handler (\(_ :: IOException) -> reconnectRetry robustConnection) -- Redis unreachable
```

Comment:
Can you also add logs here? Without this it is a little confusing why the lazy connection starts restarting. Also, it would be nice to know which type of errors are causing a connection restart.